lib/jetstream/pull_consumer.ex

```diff
@@ -1,34 +1,295 @@
+          optional(:deliver_policy) =>
+            :all | :last | :new | :by_start_sequence | :by_start_time | :last_per_subject,
+          optional(:deliver_subject) => binary(),
+          optional(:description) => binary(),
+          optional(:filter_subject) => binary(),
+          optional(:flow_control) => boolean(),
+          optional(:headers_only) => boolean(),
+          optional(:idle_heartbeat) => non_neg_integer(),
+          optional(:inactive_threshold) => non_neg_integer(),
+          optional(:max_ack_pending) => integer(),
+          optional(:max_batch) => integer(),
+          optional(:max_deliver) => integer(),
+          optional(:max_expires) => non_neg_integer(),
+          optional(:max_waiting) => integer(),
+          optional(:opt_start_seq) => non_neg_integer(),
+          optional(:opt_start_time) => DateTime.t(),
+          optional(:rate_limit_bps) => non_neg_integer(),
+          optional(:replay_policy) => :instant | :original,
+          optional(:sample_freq) => binary()
+        }
+
+  defmacro __using__(_opts) do
+    quote do
+      @behaviour Jetstream.PullConsumer
+
+      @doc """
+      Returns a specification to start this module under a supervisor.
+      See `Supervisor`.
+      """
+      @spec child_spec(
+              init_arg :: %{
+                connection_name: pid() | atom(),
+                stream: binary() | Jetstream.API.Stream.t(),
+                consumer: binary() | Jetstream.API.Consumer.t()
+              }
+            ) :: Supervisor.child_spec()
+      def child_spec(init_arg) do
+        Jetstream.PullConsumer.child_spec(__MODULE__, init_arg)
+      end
+
+      defoverridable child_spec: 1
+    end
+  end
+
+  @spec start_link(module :: module(), settings :: settings(), options :: GenServer.options()) ::
+          GenServer.on_start()
+  def start_link(module, settings, options \\ []) do
+    settings =
+      settings
+      |> Map.put(:module, module)
+
+    GenServer.start_link(__MODULE__, settings, options)
+  end
+
+  @spec child_spec(module :: module(), init_arg :: settings()) :: Supervisor.child_spec()
+  def child_spec(module, init_arg) do
+    %{
+      id: __MODULE__,
+      start:
+        {__MODULE__, :start_link,
+         [
+           module,
+           init_arg
+         ]}
     }
+  end
+
+  def init(arg) do
+    {:ok, arg, {:continue, :connect}}
+  end
+
+  def handle_continue(:connect, %{
+        connection_name: connection_name,
+        stream: stream,
+        consumer: consumer,
+        module: module
+      }) do
+    with {:ok, conn} <- connection_pid(connection_name),
+         {:ok, _response} <- create_stream(conn, stream),
+         {:ok, _response} <- create_consumer(conn, consumer),
+         listening_topic = "_CON.#{nuid()}",
+         {:ok, _sid} <- Gnat.sub(conn, self(), listening_topic),
+         :ok <- next_message(conn, stream_name(stream), consumer_name(consumer), listening_topic) do
+      Process.link(conn)
```

Comment:
I think we should `link` **right after** finding connection PID. `Gnat.sub` and `next_message` calls are IO-bound and I can imagine a situation where connection fails during these calls.